File: zipkin-server/src/test/java/zipkin2/server/internal/throttle/ThrottledCallTest.java
/*
 * Copyright The OpenZipkin Authors
 * SPDX-License-Identifier: Apache-2.0
 */
package zipkin2.server.internal.throttle;

import com.linecorp.armeria.common.metric.NoopMeterRegistry;
import com.netflix.concurrency.limits.Limiter;
import com.netflix.concurrency.limits.Limiter.Listener;
import com.netflix.concurrency.limits.limit.SettableLimit;
import com.netflix.concurrency.limits.limiter.SimpleLimiter;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Optional;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Predicate;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import zipkin2.Call;
import zipkin2.Callback;

import static com.linecorp.armeria.common.util.Exceptions.clearTrace;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.failBecauseExceptionWasNotThrown;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static zipkin2.server.internal.throttle.ThrottledCall.NOOP_CALLBACK;
import static zipkin2.server.internal.throttle.ThrottledCall.STORAGE_THROTTLE_MAX_CONCURRENCY;
import static zipkin2.server.internal.throttle.ThrottledStorageComponent.STORAGE_THROTTLE_MAX_QUEUE_SIZE;

class ThrottledCallTest {
  SettableLimit limit = SettableLimit.startingAt(0);
  SimpleLimiter limiter = SimpleLimiter.newBuilder().limit(limit).build();
  LimiterMetrics limiterMetrics = new LimiterMetrics(NoopMeterRegistry.get());
  Predicate<Throwable> isOverCapacity = RejectedExecutionException.class::isInstance;

  int numThreads = 1;
  ExecutorService executor = Executors.newSingleThreadExecutor();

  @AfterEach void shutdownExecutor() {
    executor.shutdown();
  }

  @Test void niceToString() {
    Call<Void> delegate = mock(Call.class);
    when(delegate.toString()).thenReturn("StoreSpansCall{}");

    assertThat(new ThrottledCall(delegate, executor, limiter, limiterMetrics, isOverCapacity))
      .hasToString("Throttled(StoreSpansCall{})");
  }

  @Test void execute_isThrottled() throws Exception {
    int queueSize = 1;
    int totalTasks = numThreads + queueSize;
    limit.setLimit(totalTasks);

    Semaphore startLock = new Semaphore(numThreads);
    Semaphore waitLock = new Semaphore(totalTasks);
    Semaphore failLock = new Semaphore(1);
    ThrottledCall throttled = throttle(new LockedCall(startLock, waitLock));

    // Step 1: drain appropriate locks
    startLock.drainPermits();
    waitLock.drainPermits();
    failLock.drainPermits();

    // Step 2: saturate threads and fill queue
    ExecutorService backgroundPool = Executors.newCachedThreadPool();
    for (int i = 0; i < totalTasks; i++) {
      backgroundPool.submit(() -> throttled.clone().execute());
    }

    try {
      // Step 3: make sure the threads actually started
      startLock.acquire(numThreads);

      // Step 4: submit something beyond our limits
      Future<?> future = backgroundPool.submit(() -> {
        try {
          throttled.execute();
        } catch (IOException e) {
          throw new UncheckedIOException(e);
        } finally {
          // Step 6: signal that we tripped the limit
          failLock.release();
        }
      });

      // Step 5: wait to make sure our limit actually tripped
      failLock.acquire();

      future.get();

      // Step 7: Expect great things
      failBecauseExceptionWasNotThrown(ExecutionException.class);
    } catch (ExecutionException t) {
      assertThat(t)
        .isInstanceOf(ExecutionException.class) // from future.get
        .hasCauseInstanceOf(RejectedExecutionException.class);
    } finally {
      waitLock.release(totalTasks);
      startLock.release(totalTasks);
      backgroundPool.shutdownNow();
    }
  }

  @Test void execute_throttlesBack_whenStorageRejects() throws Exception {
    Listener listener = mock(Listener.class);
    FakeCall call = new FakeCall();
    call.overCapacity = true;

    ThrottledCall throttle =
      new ThrottledCall(call, executor, mockLimiter(listener), limiterMetrics, isOverCapacity);

    try {
      throttle.execute();
      assertThat(true).isFalse(); // should raise a RejectedExecutionException
    } catch (RejectedExecutionException e) {
      verify(listener).onDropped();
    }
  }

  @Test void execute_ignoresLimit_whenPoolFull() throws Exception {
    Listener listener = mock(Listener.class);

    ThrottledCall throttle = new ThrottledCall(new FakeCall(), mockExhaustedPool(),
      mockLimiter(listener), limiterMetrics, isOverCapacity);

    try {
      throttle.execute();
      assertThat(true).isFalse(); // should raise a RejectedExecutionException
    } catch (RejectedExecutionException e) {
      verify(listener).onIgnore();
    }
  }

  @Test void enqueue_isThrottled() throws Exception {
    int queueSize = 1;
    int totalTasks = numThreads + queueSize;
    limit.setLimit(totalTasks);

    Semaphore startLock = new Semaphore(numThreads);
    Semaphore waitLock = new Semaphore(totalTasks);
    ThrottledCall throttle = throttle(new LockedCall(startLock, waitLock));

    // Step 1: drain appropriate locks
    startLock.drainPermits();
    waitLock.drainPermits();

    // Step 2: saturate threads and fill queue
    Callback<Void> callback = mock(Callback.class);
    for (int i = 0; i < totalTasks; i++) {
      throttle.clone().enqueue(callback);
    }

    // Step 3: make sure the threads actually started
    startLock.acquire(numThreads);

    // Step 4: submit something beyond our limits and make sure it fails
    assertThatThrownBy(() -> throttle.clone().enqueue(callback))
      .isEqualTo(STORAGE_THROTTLE_MAX_CONCURRENCY);
  }

  @Test void enqueue_throttlesBack_whenStorageRejects() throws Exception {
    Listener listener = mock(Listener.class);
    FakeCall call = new FakeCall();
    call.overCapacity = true;

    ThrottledCall throttle =
      new ThrottledCall(call, executor, mockLimiter(listener), limiterMetrics, isOverCapacity);

    final CountDownLatch countDown = new CountDownLatch(1);
    final AtomicReference<Throwable> throwable = new AtomicReference<>();
    throttle.enqueue(new Callback<>() {
      @Override public void onSuccess(Void value) {
        countDown.countDown();
      }

      @Override public void onError(Throwable t) {
        throwable.set(t);
        countDown.countDown();
      }
    });

    countDown.await();

    assertThat(throwable).hasValue(OVER_CAPACITY);

    verify(listener).onDropped();
  }

  @Test void enqueue_ignoresLimit_whenPoolFull() {
    Listener listener = mock(Listener.class);

    ThrottledCall throttle = new ThrottledCall(new FakeCall(), mockExhaustedPool(),
      mockLimiter(listener), limiterMetrics, isOverCapacity);

    assertThatThrownBy(() -> throttle.enqueue(NOOP_CALLBACK))
      .isEqualTo(STORAGE_THROTTLE_MAX_QUEUE_SIZE);

    verify(listener).onIgnore();
  }

  ThrottledCall throttle(Call<Void> delegate) {
    return new ThrottledCall(delegate, executor, limiter, limiterMetrics, isOverCapacity);
  }

  static final class LockedCall extends Call.Base<Void> {
    final Semaphore startLock, waitLock;

    LockedCall(Semaphore startLock, Semaphore waitLock) {
      this.startLock = startLock;
      this.waitLock = waitLock;
    }

    @Override public Void doExecute() {
      try {
        startLock.release();
        waitLock.acquire();
        return null;
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new AssertionError(e);
      }
    }

    @Override public void doEnqueue(Callback<Void> callback) {
      try {
        callback.onSuccess(doExecute());
      } catch (Throwable t) {
        propagateIfFatal(t);
        callback.onError(t);
      }
    }

    @Override public LockedCall clone() {
      return new LockedCall(startLock, waitLock);
    }
  }

  ExecutorService mockExhaustedPool() {
    ExecutorService mock = mock(ExecutorService.class);
    doThrow(STORAGE_THROTTLE_MAX_QUEUE_SIZE).when(mock).execute(any());
    doThrow(STORAGE_THROTTLE_MAX_QUEUE_SIZE).when(mock).submit(any(Callable.class));
    return mock;
  }

  Limiter<Void> mockLimiter(Listener listener) {
    Limiter<Void> mock = mock(Limiter.class);
    when(mock.acquire(any())).thenReturn(Optional.of(listener));
    return mock;
  }

  static final Exception OVER_CAPACITY = clearTrace(new RejectedExecutionException("overCapacity"));

  static final class FakeCall extends Call.Base<Void> {
    boolean overCapacity = false;

    @Override public Void doExecute() {
      throw new AssertionError("throttling never uses execute");
    }

    @Override public void doEnqueue(Callback<Void> callback) {
      if (overCapacity) {
        callback.onError(OVER_CAPACITY);
      } else {
        callback.onSuccess(null);
      }
    }

    @Override public FakeCall clone() {
      return new FakeCall();
    }
  }
}

File: zipkin-server/src/test/kotlin/zipkin2/server/internal/throttle/ThrottledCallTest.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package zipkin2.server.internal.throttle

import com.netflix.concurrency.limits.Limiter
import com.netflix.concurrency.limits.Limiter.Listener
import com.netflix.concurrency.limits.limit.SettableLimit
import com.netflix.concurrency.limits.limiter.SimpleLimiter
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.mockito.Mockito.doThrow
import org.mockito.Mockito.verify
import zipkin2.Call
import zipkin2.Callback
import java.io.IOException
import java.util.Optional
import java.util.concurrent.Callable
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.Semaphore
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class ThrottledCallTest {
  val limit = SettableLimit.startingAt(0)
  val limiter = SimpleLimiter.newBuilder().limit(limit).build<Void>()

  val numThreads = 1
  val executor = Executors.newSingleThreadExecutor();
  @After fun shutdownExecutor() = executor.shutdown()

  inline fun <reified T : Any> mock() = Mockito.mock(T::class.java)

  @Test fun niceToString() {
    val delegate: Call<Void> = mock()
    `when`(delegate.toString()).thenReturn("StoreSpansCall{}")

    assertThat(ThrottledCall(executor, limiter, delegate))
      .hasToString("Throttled(StoreSpansCall{})")
  }

  @Test fun execute_isThrottled() {
    val queueSize = 1
    val totalTasks = numThreads + queueSize
    limit.limit = totalTasks

    val startLock = Semaphore(numThreads)
    val waitLock = Semaphore(totalTasks)
    val failLock = Semaphore(1)
    val throttled = throttle(LockedCall(startLock, waitLock))

    // Step 1: drain appropriate locks
    startLock.drainPermits()
    waitLock.drainPermits()
    failLock.drainPermits()

    // Step 2: saturate threads and fill queue
    val backgroundPool = Executors.newCachedThreadPool()
    for (i in 0 until totalTasks) {
      backgroundPool.submit(Callable { throttled.clone().execute() })
    }

    try {
      // Step 3: make sure the threads actually started
      startLock.acquire(numThreads)

      // Step 4: submit something beyond our limits
      val future = backgroundPool.submit(Callable {
        try {
          throttled.execute()
        } catch (e: IOException) {
          throw RuntimeException(e)
        } finally {
          // Step 6: signal that we tripped the limit
          failLock.release()
        }
      })

      // Step 5: wait to make sure our limit actually tripped
      failLock.acquire()

      future.get()

      // Step 7: Expect great things
      assertThat(true).isFalse() // should raise a RejectedExecutionException
    } catch (t: Throwable) {
      assertThat(t)
        .isInstanceOf(ExecutionException::class.java) // from future.get
        .hasCauseInstanceOf(RejectedExecutionException::class.java)
    } finally {
      waitLock.release(totalTasks)
      startLock.release(totalTasks)
      backgroundPool.shutdownNow()
    }
  }

  @Test fun execute_trottlesBack_whenStorageRejects() {
    val listener: Listener = mock()
    val call = FakeCall()
    call.overCapacity = true

    val throttle = ThrottledCall(executor, mockLimiter(listener), call)
    try {
      throttle.execute()
      assertThat(true).isFalse() // should raise a RejectedExecutionException
    } catch (e: RejectedExecutionException) {
      verify(listener).onDropped()
    }
  }

  @Test fun execute_ignoresLimit_whenPoolFull() {
    val listener: Listener = mock()

    val throttle = ThrottledCall(mockExhaustedPool(), mockLimiter(listener), FakeCall())
    try {
      throttle.execute()
      assertThat(true).isFalse() // should raise a RejectedExecutionException
    } catch (e: RejectedExecutionException) {
      verify(listener).onIgnore()
    }
  }

  @Test fun enqueue_isThrottled() {
    val queueSize = 1
    val totalTasks = numThreads + queueSize
    limit.limit = totalTasks

    val startLock = Semaphore(numThreads)
    val waitLock = Semaphore(totalTasks)
    val throttle = throttle(LockedCall(startLock, waitLock))

    // Step 1: drain appropriate locks
    startLock.drainPermits()
    waitLock.drainPermits()

    // Step 2: saturate threads and fill queue
    val callback: Callback<Void> = mock()
    for (i in 0 until totalTasks) {
      throttle.clone().enqueue(callback)
    }

    // Step 3: make sure the threads actually started
    startLock.acquire(numThreads)

    try {
      // Step 4: submit something beyond our limits and make sure it fails
      throttle.clone().enqueue(callback)

      assertThat(true).isFalse() // should raise a RejectedExecutionException
    } catch (e: RejectedExecutionException) {
    } finally {
      waitLock.release(totalTasks)
      startLock.release(totalTasks)
    }
  }

  @Test fun enqueue_throttlesBack_whenStorageRejects() {
    val listener: Listener = mock()
    val call = FakeCall()
    call.overCapacity = true

    val throttle = ThrottledCall(executor, mockLimiter(listener), call)
    val latch = CountDownLatch(1)
    throttle.enqueue(object : Callback<Void> {
      override fun onSuccess(value: Void) {
        latch.countDown()
      }

      override fun onError(t: Throwable) {
        latch.countDown()
      }
    })

    latch.await(1, TimeUnit.MINUTES)
    verify(listener).onDropped()
  }

  @Test fun enqueue_ignoresLimit_whenPoolFull() {
    val listener: Listener = mock()

    val throttle =
      ThrottledCall(mockExhaustedPool(), mockLimiter(listener), FakeCall())
    try {
      throttle.enqueue(null)
      assertThat(true).isFalse() // should raise a RejectedExecutionException
    } catch (e: RejectedExecutionException) {
      verify(listener).onIgnore()
    }
  }

  private fun throttle(delegate: Call<Void>) = ThrottledCall(executor, limiter, delegate)

  private class LockedCall(val startLock: Semaphore, val waitLock: Semaphore) : Call.Base<Void>() {
    override fun doExecute(): Void? {
      try {
        startLock.release()
        waitLock.acquire()
        return null;
      } catch (e: InterruptedException) {
        Thread.currentThread().interrupt()
        throw AssertionError(e)
      }
    }

    override fun doEnqueue(callback: Callback<Void>) {
      try {
        callback.onSuccess(doExecute())
      } catch (t: Throwable) {
        callback.onError(t)
      }
    }

    override fun clone() = LockedCall(startLock, waitLock);
  }

  private fun mockExhaustedPool(): ExecutorService {
    val mock: ExecutorService = mock()
    doThrow(RejectedExecutionException::class.java).`when`(mock).execute(any())
    doThrow(RejectedExecutionException::class.java).`when`(mock).submit(any(Callable::class.java))
    return mock
  }

  private fun mockLimiter(listener: Listener): Limiter<Void> {
    val mock: Limiter<Void> = mock()
    `when`(mock.acquire(any())).thenReturn(Optional.of(listener))
    return mock
  }

  private class FakeCall(var overCapacity: Boolean = false) : Call.Base<Void>() {
    override fun doExecute(): Void? {
      if (overCapacity) throw RejectedExecutionException()
      return null
    }

    override fun doEnqueue(callback: Callback<Void>) {
      if (overCapacity) {
        callback.onError(RejectedExecutionException())
      } else {
        callback.onSuccess(null)
      }
    }

    override fun clone() = FakeCall()
  }
}
